File: frontend/apps/ios/Pantopus/Features/Mailbox/ItemDetail/Bodies/Components/CertifiedTermsSheet.swift
//
//  CertifiedTermsSheet.swift
//  Pantopus
//
//  Modal sheet shown when the user taps "View terms" on a certified
//  mail item. Renders the terms URL with a primary CTA that opens it
//  in the system browser. Fetching the document body is left for a
//  later prompt — the design says "fetch as plain text or render as
//  in-line markdown if backend returns markdown" but no fetch endpoint
//  exists today.
//

import SwiftUI

@MainActor
public struct CertifiedTermsSheet: View {
    private let termsURL: URL
    private let onDismiss: @MainActor () -> Void
    @Environment(\.openURL) private var openURL

    public init(termsURL: URL, onDismiss: @escaping @MainActor () -> Void) {
        self.termsURL = termsURL
        self.onDismiss = onDismiss
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s4) {
            HStack {
                Text("Terms of certified delivery")
                    .pantopusTextStyle(.h3)
                    .foregroundStyle(Theme.Color.appText)
                    .accessibilityAddTraits(.isHeader)
                Spacer()
                Button(action: { onDismiss() }) {
                    Icon(.x, size: 20, color: Theme.Color.appText)
                        .frame(width: 44, height: 44)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Close")
            }

            Text("Read the full terms before acknowledging this certified document.")
                .pantopusTextStyle(.small)
                .foregroundStyle(Theme.Color.appTextSecondary)

            // TODO(content): fetch body text from `termsURL` and render
            // inline once a backend endpoint exposes the document. For
            // now we surface the URL itself so the user can open it.
            VStack(alignment: .leading, spacing: Spacing.s2) {
                Text("Document URL")
                    .pantopusTextStyle(.caption)
                    .foregroundStyle(Theme.Color.appTextSecondary)
                Text(termsURL.absoluteString)
                    .pantopusTextStyle(.small)
                    .foregroundStyle(Theme.Color.appText)
                    .lineLimit(3)
                    .truncationMode(.middle)
            }
            .padding(Spacing.s3)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Theme.Color.appSurfaceSunken)
            .clipShape(RoundedRectangle(cornerRadius: Radii.md))

            PrimaryButton(title: "Open in browser") {
                await MainActor.run {
                    openURL(termsURL)
                    onDismiss()
                }
            }

            Spacer(minLength: 0)
        }
        .padding(Spacing.s4)
        .background(Theme.Color.appBg)
        .presentationDetents([.medium, .large])
        .presentationDragIndicator(.visible)
        .accessibilityIdentifier("certifiedTermsSheet")
    }
}

#Preview {
    CertifiedTermsSheet(
        termsURL: URL(string: "https://example.com/certified-terms.pdf")!,
        onDismiss: {}
    )
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/mailbox/item_detail/bodies/components/CertifiedTermsSheet.kt
@file:Suppress("MagicNumber", "PackageNaming")

package app.pantopus.android.ui.screens.mailbox.item_detail.bodies.components

import android.content.Intent
import android.net.Uri
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.style.TextOverflow
import app.pantopus.android.ui.components.PrimaryButton
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusTextStyle
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/** Test tag on the certified-mail terms sheet. */
const val CERTIFIED_TERMS_SHEET_TAG = "certifiedTermsSheet"

/**
 * Modal bottom sheet shown when the user taps "View terms" on a
 * certified mail item. Surfaces the [termsUrl] with a primary CTA
 * that opens it in the system browser. Fetching the document body is
 * left for a later prompt — the design says "fetch as plain text or
 * render as in-line markdown" but no fetch endpoint exists today.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CertifiedTermsSheet(
    termsUrl: String,
    onDismiss: () -> Unit,
) {
    val context = LocalContext.current
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false)
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = PantopusColors.appBg,
        modifier = Modifier.testTag(CERTIFIED_TERMS_SHEET_TAG),
    ) {
        Column(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .padding(Spacing.s4),
            verticalArrangement = Arrangement.spacedBy(Spacing.s3),
        ) {
            Text(
                text = "Terms of certified delivery",
                style = PantopusTextStyle.h3,
                color = PantopusColors.appText,
            )
            Text(
                text = "Read the full terms before acknowledging this certified document.",
                style = PantopusTextStyle.small,
                color = PantopusColors.appTextSecondary,
            )
            Column(
                modifier =
                    Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(Radii.md))
                        .background(PantopusColors.appSurfaceSunken)
                        .padding(Spacing.s3),
                verticalArrangement = Arrangement.spacedBy(Spacing.s1),
            ) {
                Text(
                    text = "Document URL",
                    style = PantopusTextStyle.caption,
                    color = PantopusColors.appTextSecondary,
                )
                Text(
                    text = termsUrl,
                    style = PantopusTextStyle.small,
                    color = PantopusColors.appText,
                    maxLines = 3,
                    overflow = TextOverflow.Ellipsis,
                )
            }
            PrimaryButton(
                title = "Open in browser",
                onClick = {
                    val intent =
                        Intent(Intent.ACTION_VIEW, Uri.parse(termsUrl)).apply {
                            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        }
                    runCatching { context.startActivity(intent) }
                    onDismiss()
                },
                modifier = Modifier.fillMaxWidth(),
            )
        }
    }
}
